### ISSUE TITLE
clang: incorrectly accepts addition with an array prvalue operand

### ISSUE THREAD

**Issue by languagelawyer** [llvm/llvm-project#54016]
```c++
int main()
{
    using IA = int[];

    IA{ 1, 2, 3 } + 0;
}
```
[expr.add]/1:
> For addition, either both operands shall have arithmetic or unscoped enumeration type, or one operand shall be a pointer to a completely-defined object type and the other shall have integral or unscoped enumeration type.

Array types are not allowed. Neither there is wording requiring the array-to-pointer conversion for an operand.

**Comment by shafik** [comment]
I believe clang is correct here, we apply a standard conversion [\[conv.general\]](http://eel.is/c++draft/conv#general-1.1) which includes array-to-pointer conversion [\[conv.array\]](http://eel.is/c++draft/conv#array) which results in a temporary materialization conversion [\[conv.rval\]](http://eel.is/c++draft/conv#rval).

A quick [godbolt](https://godbolt.org/z/1W9rs5rP1) with AST output seems to confirm this what clang is doing.

This [2018 cfe-dev mailing](https://lists.llvm.org/pipermail/cfe-dev/2018-February/056928.html) on a related case seems to indicate that clang is correct and although "GCC's behaviour is not unreasonable" CC @zygoloid 

Clang has -Waddress-of-temporary but looking at the test suite in:

```
  clang/test/SemaCXX/address-of-temporary.cpp
```

It looks like we purposely do not consider the array-to-pointer conversion to be included as taking the address of a temporary.

Although I can see the case can be made for clang to be stricter. 

Note, `-Wdangling` will catch the case where we pointer to the temporary outlives the temporary.

**Comment by languagelawyer** [comment]
> I believe clang is correct here, we apply a standard conversion [[conv.general]](http://eel.is/c++draft/conv#general-1.1)

And why it should be applied here?

**Comment by zygoloid** [comment]
> > I believe clang is correct here, we apply a standard conversion [[conv.general]](http://eel.is/c++draft/conv#general-1.1)
> 
> And why it should be applied here?

I see, your argument is that [[basic.lval]/6](http://eel.is/c++draft/basic.lval#6) does not apply because the array operand is not a glvalue? I don't think that's the intent of CWG1232, but it does seem to be what the current wording says. This seems worth a CWG discussion before we proceed.

**Comment by languagelawyer** [comment]
> your argument is that [[basic.lval]/6](http://eel.is/c++draft/basic.lval#6) does not apply because the array operand is not a glvalue?

Exactly.

> This seems worth a CWG discussion

e-mail thread or immediate issue request?

**Comment by zygoloid** [comment]
I think it's OK to go straight to filing an issue (github.com/cplusplus/cwg/issues), given that I think it's at least unclear whether the wording reflects the intent.

**Comment by shafik** [comment]
Also would like to point out [\[conv.general\]p2.1](http://eel.is/c++draft/conv#general-2.1) says:

>Expressions with a given type will be implicitly converted to other types in several contexts:
> - When used as operands of operators[.](http://eel.is/c++draft/conv#general-2.1.sentence-1)
The operator's requirements for its operands dictate the destination type ([[expr.compound]](http://eel.is/c++draft/expr.compound))[.](http://eel.is/c++draft/conv#general-2.1.sentence-2)

Although this is a note and this non-normative I think it speaks to the intent.

**Comment by jensmaurer** [comment]
Created [core issue 2548](https://cplusplus.github.io/CWG/issues/2548.html).

**Comment by languagelawyer** [comment]
> > * The operator's requirements for its operands dictate the destination type ([[expr.compound]](http://eel.is/c++draft/expr.compound))[.](http://eel.is/c++draft/conv#general-2.1.sentence-2)
> 
> Although this is a note and this non-normative I think it speaks to the intent.

No, it is just a defective Note. See https://timsong-cpp.github.io/cppwp/n4868/dcl.init.general#def:destination_type. Operand of a built-in operator is not «object or reference being initialized».

**Comment by pinskia** [comment]
Note the GCC issue side is https://gcc.gnu.org/bugzilla/show_bug.cgi?id=94264 .

**Comment by languagelawyer** [comment]
@pinskia GCC has no issue

**Comment by zygoloid** [comment]
CWG forwarded the issue to EWG: https://github.com/cplusplus/papers/issues/1633

There's no clear indication yet of whether prvalue arrays should be subject to the array-to-pointer conversion when they appear as the operand of `[]` or unary `*`.

**Comment by languagelawyer** [comment]
> when they appear as the operand of `[]` or unary `*`.

You mean `+` instead of `[]`?

**Comment by Fedr** [comment]
Related discussion: https://stackoverflow.com/q/77051011/7325599

**Comment by zygoloid** [comment]
> You mean `+` instead of `[]`?

I meant `[]`, as I wrote, but unary `+` is another operator for which the same concerns apply. (I mentioned this to Jens yesterday and the CWG issue has already been updated to cover that case.)

**Comment by Eisenwave** [comment]
After the resolution of CWG2548, it is clear that `+` for prvalue array operands is not valid, and this is intentional. @zygoloid can you mark this as confirmed?

**Comment by languagelawyer** [comment]
> After the resolution of CWG2548, it is clear that `+` and `*` for prvalue array operands is not valid, and this is intentional

It seems they, against EWG direction, made it valid in P2900R14, namely, by applying temporary materialization conversion to almost all prvalues http://eel.is/c++draft/basic.lval#7:
> Unless otherwise specified …, whenever a prvalue that is not the result of the lvalue-to-rvalue conversion ([[conv.lval]](http://eel.is/c++draft/conv.lval)) appears as an operand of an operator, the [temporary materialization conversion](http://eel.is/c++draft/conv.rval) is applied to convert the expression to an xvalue
